src/middlewared/middlewared/plugins/smart.py

```diff
@@ -21,159 +22,142 @@
 RE_TIME = re.compile(r'test will complete after ([a-z]{3} [a-z]{3} [0-9 ]+ \d\d:\d\d:\d\d \d{4})', re.IGNORECASE)
 RE_TIME_SCSIPRINT_EXTENDED = re.compile(r'Please wait (\d+) minutes for test to complete')
 
-RE_OF_TEST_REMAINING = re.compile(r'([0-9]+)% of test remaining')
-RE_SELF_TEST_STATUS = re.compile(r'self-test in progress \(([0-9]+)% completed\)')
-
 
 async def annotate_disk_smart_tests(middleware, tests_filter, disk):
     if disk["disk"] is None:
         return
 
-    output = await middleware.call("disk.smartctl", disk["disk"], ["-a"], {"silent": True})
+    output = await middleware.call("disk.smartctl", disk["disk"], ["-a", "--json=c"], {"silent": True})
     if output is None:
         return
+    data = json.loads(output)
 
-    tests = parse_smart_selftest_results(output) or []
-    current_test = parse_current_smart_selftest(output)
+    tests = parse_smart_selftest_results(data) or []
+    current_test = parse_current_smart_selftest(data)
     return dict(tests=filter_list(tests, tests_filter), current_test=current_test, **disk)
 
 
-def parse_smart_selftest_results(stdout):
+def parse_smart_selftest_results(data):
     tests = []
 
     # ataprint.cpp
-    if "LBA_of_first_error" in stdout:
-        for line in stdout.split("\n"):
-            if not line.startswith("#"):
-                continue
-
-            if line[58] == "%":
-                remaining = line[55:58]
-                lifetime = line[61:69]
-            else:
-                remaining = line[55:57]
-                lifetime = line[60:68]
+    if "ata_smart_self_test_log" in data:
+        if "table" in data["ata_smart_self_test_log"]["standard"]: # If there are no tests, there is no table
+            for index, entry in enumerate(data["ata_smart_self_test_log"]["standard"]["table"]):
 
-            test = {
-                "num": int(line[1:3].strip()),
-                "description": line[5:24].strip(),
-                "status_verbose": line[25:54].strip(),
-                "remaining": int(remaining.strip()) / 100,
-                "lifetime": int(lifetime.strip()),
-                "lba_of_first_error": line[77:].strip(),
-            }
+                # remaining_percent is in the dict only if the test is in progress (status value & 0x0f)
+                if remaining := entry["status"]["value"] & 0x0f:
```

Comment:
Can we have this as a constant that you would copy from smartmontools source? Or is this a hex literal there also?

Reply:
It's also a hex literal in the source, the last byte has the progress encoded in it.